lib/hooks/userspace/hook-grpc.js

```diff
@@ -129,6 +128,299 @@ function makeClientConstructorWrap(makeClientConstructor) {
+          // If there is an error, then no result will be serialized to
+          // be sent back to the client. So end the root span here.
+          endRootSpanForRequest(rootContext, requestName);
+        }
+        return callback(err, result);
+      };
+      return func.apply(this, arguments);
+    }, context);
+  });
+  shimmer.wrap(handlerSet, 'serialize', function (serialize) {
+    return function serializeTrace(result) {
+      var serializeResult = serialize.apply(this, arguments);
+      endRootSpanForRequest(rootContext, requestName);
+      return serializeResult;
+    };
+  });
+}
+
+/**
+ * Wraps a unary function in order to record trace spans.
+ * @param {cls.Namespace} namespace The CLS namespace.
+ * @param {Object} handlerSet An object containing references to the function handle,
+ * as well as serialize and deserialize handles.
+ * @param {string} requestName The human-friendly name of the request.
+ */
+function wrapUnary(namespace, handlerSet, requestName) {
+  var rootContext;
+  // CLS context object
+  var context = {};
+  // Start right before deserialization
+  // End right after serialization
+  shimmer.wrap(handlerSet, 'deserialize', function (deserialize) {
+    return namespace.bind(function deserializeTrace() {
+      rootContext = startRootSpanForRequest(requestName);
+      return deserialize.apply(this, arguments);
+    }, context);
+  });
+  // Even though our root span doesn't start or end here,
+  // we need to wrap 'func' so that:
+  // (1) child spans have the correct context, and
+  // (2) we can get additional labels as needed
+  shimmer.wrap(handlerSet, 'func', function (func) {
+    return namespace.bind(function funcTrace(call, callback) {
+      if (agent.config_.enhancedDatabaseReporting) {
+        shimmer.wrap(call, 'sendMetadata', sendMetadataWrapper(rootContext));
+      }
+      if (agent.config_.enhancedDatabaseReporting) {
+        rootContext.addLabel('argument', JSON.stringify(call.request));
+      }
+      // arguments[1] is the callback
+      arguments[1] = function (err, result, trailer) {
+        if (agent.config_.enhancedDatabaseReporting) {
+          if (err) {
+            rootContext.addLabel('error', err);
+          }
+          else {
+            rootContext.addLabel('result', JSON.stringify(result));
+          }
+          if (trailer) {
+            rootContext.addLabel('trailing_metadata', JSON.stringify(trailer.getMap()));
+          }
+        }
+        if (err) {
+          // If there is an error, then no result will be serialized to
+          // be sent back to the client. So end the root span here.
+          endRootSpanForRequest(rootContext, requestName);
+        }
+        return callback(err, result);
+      };
+      return func.apply(this, arguments);
+    }, context);
+  });
+  shimmer.wrap(handlerSet, 'serialize', function (serialize) {
+    return function serializeTrace() {
+      var serializeResult = serialize.apply(this, arguments);
```

Comment:
That is a good point - yes, both the docs and usage of serialize suggest that it's synchronous.